packages/upload-api/src/index/add.js

```diff
@@ -0,0 +1,75 @@
+import * as Server from '@ucanto/server'
+import { ok, error } from '@ucanto/server'
+import * as Index from '@web3-storage/capabilities/index'
+import * as ShardedDAGIndex from './lib/sharded-dag-index.js'
+import * as API from '../types.js'
+
+/**
+ * @param {API.IndexServiceContext} context
+ * @returns {API.ServiceMethod<API.IndexAdd, API.IndexAddSuccess, API.IndexAddFailure>}
+ */
+export const provide = (context) =>
+  Server.provide(Index.add, (input) => add(input, context))
+
+/**
+ * @param {API.Input<Index.add>} input
+ * @param {API.IndexServiceContext} context
+ * @returns {Promise<API.Result<API.IndexAddSuccess, API.IndexAddFailure>>}
+ */
+const add = async ({ capability }, context) => {
+  const space = capability.with
+  const idxLink = capability.nb.index
+
+  // ensure the index was stored in the agent's space
+  const idxAllocRes = await assertAllocated(
+    context,
+    space,
+    idxLink.multihash,
+    'IndexNotFound'
+  )
+  if (!idxAllocRes.ok) return idxAllocRes
+
+  // fetch the index from the network
+  const idxBlobRes = await context.blobRetriever.stream(idxLink.multihash)
+  if (!idxBlobRes.ok) {
+    if (idxBlobRes.error.name === 'BlobNotFound') {
+      return error(
+        /** @type {API.IndexNotFound} */
+        ({ name: 'IndexNotFound', digest: idxLink.multihash.bytes })
```

Comment:
Shouldn't we keep the `BlobNotFound` error to differentiate between when an index does not exist in the set of blob indexes (IndexNotFound) from when the index exists but what it points to cannot be found (`BlobNotFound`)?

Reply:
Yes, exactly, the index is a blob also, this is just differentiating between generic blob not being found and something that says the actual index was not found.